sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -147,43 +172,59 @@ private void drainQueue() {
         long numberRequested = REQUESTED.get(this);
         boolean isEmpty = bufferMessages.isEmpty();
 
-        SynchronousReceiveWork work;
+        SynchronousReceiveWork currentDownstream = null;
         while (numberRequested != 0L && !isEmpty) {
             if (isTerminated()) {
                 break;
             }
 
-            long numberEmitted = 0L;
-            while (numberRequested != numberEmitted) {
+            // Track the number of messages read from the buffer those are either emitted to downstream or released.
+            long numberConsumed = 0L;
+            // Iterate and attempt to read the requested number of events.
+            while (numberRequested != numberConsumed) {
                 if (isEmpty || isTerminated()) {
                     break;
                 }
 
                 final ServiceBusReceivedMessage message = bufferMessages.poll();
+
+                // While there are messages in the buffer, obtain the current (unterminated) downstream and
+                // attempt to emit the message to it.
                 boolean isEmitted = false;
                 while (!isEmitted) {
-                    work = getOrUpdateCurrentWork();
-                    if (work == null) {
+                    currentDownstream = getOrUpdateCurrentWork();
+                    if (currentDownstream == null) {
                         break;
                     }
 
-                    isEmitted = work.emitNext(message);
+                    isEmitted = currentDownstream.emitNext(message);
                 }
 
-                // We could not emit the last message that we polled because there were no work items.
-                // Push this back to the head of the work queue.
                 if (!isEmitted) {
-                    bufferMessages.addFirst(message);
-                    break;
+                    // The only reason we can't emit was the downstream(s) are/were terminated hence nobody
+                    // to receive the message.
+                    if (isPrefetchDisabled) {
+                        // release is enabled only for no-prefetch scenario.
+                        asyncClient.release(message).subscribe(__ -> { },
+                            error -> logger.warning("lockToken[{}] Couldn't release the message.",
+                                message.getLockToken(), error),
+                            () -> logger.verbose("lockToken[{}] Message successfully released.",
+                                message.getLockToken()));
+                    } else {
+                        // Re-buffer the message as it couldn't be emitted or release was disabled.
+                        bufferMessages.addFirst(message);
+                        break;
+                    }
                 }
 
-                numberEmitted++;
+                // account for consumed message - message is either emitted or released.
+                numberConsumed++;
```

Comment:
If prefetch is enabled and the message was not emitted, we should not increment this right?

Reply:
Exactly, so if the prefetch is enabled and drain couldn't emit downstream, we break from the while-loop in line 216; that way, we don't consider the message as consumed.